examples/DataApi-Voting-poll/template/modules/l0/src/main/scala/com/my/currency/l0/Main.scala

```diff
@@ -2,55 +2,75 @@ package com.my.currency.l0
+    override def genesis: DataState[VoteStateOnChain, VoteCalculatedState] = DataState(VoteStateOnChain(List.empty), VoteCalculatedState(Map.empty))
 
-      override def validateData(oldState: State, updates: NonEmptyList[Signed[PollUpdate]])(implicit context: L0NodeContext[IO]): IO[DataApplicationValidationErrorOr[Unit]] = MainData.validateData(oldState, updates)
+    override def validateUpdate(update: PollUpdate)(implicit context: L0NodeContext[IO]): IO[DataApplicationValidationErrorOr[Unit]] = ().validNec.pure[IO]
 
-      override def validateUpdate(update: PollUpdate)(implicit context: L0NodeContext[IO]): IO[DataApplicationValidationErrorOr[Unit]] = IO { ().validNec }
+    override def validateData(state: DataState[VoteStateOnChain, VoteCalculatedState], updates: NonEmptyList[Signed[PollUpdate]])(implicit context: L0NodeContext[IO]): IO[DataApplicationValidationErrorOr[Unit]] = LifecycleSharedFunctions.validateData[IO](state, updates)
 
-      override def combine(oldState: State, updates: NonEmptyList[Signed[PollUpdate]])(implicit context: L0NodeContext[IO]): IO[State] = MainData.combine(oldState, updates)
+    override def combine(state: DataState[VoteStateOnChain, VoteCalculatedState], updates: List[Signed[PollUpdate]])(implicit context: L0NodeContext[IO]): IO[DataState[VoteStateOnChain, VoteCalculatedState]] = LifecycleSharedFunctions.combine[IO](state, updates)
 
-      override def serializeState(state: State): IO[Array[Byte]] = MainData.serializeState(state)
+    override def serializeState(state: VoteStateOnChain): IO[Array[Byte]] = IO(Serializers.serializeState(state))
 
-      override def deserializeState(bytes: Array[Byte]): IO[Either[Throwable, State]] = MainData.deserializeState(bytes)
+    override def serializeUpdate(update: PollUpdate): IO[Array[Byte]] = IO(Serializers.serializeUpdate(update))
 
-      override def serializeUpdate(update: PollUpdate): IO[Array[Byte]] = MainData.serializeUpdate(update)
+    override def serializeBlock(block: Signed[DataApplicationBlock]): IO[Array[Byte]] = IO(Serializers.serializeBlock(block)(dataEncoder.asInstanceOf[Encoder[DataUpdate]]))
 
-      override def deserializeUpdate(bytes: Array[Byte]): IO[Either[Throwable, PollUpdate]] = MainData.deserializeUpdate(bytes)
+    override def deserializeState(bytes: Array[Byte]): IO[Either[Throwable, VoteStateOnChain]] = IO(Deserializers.deserializeState(bytes))
 
-      override def dataEncoder: Encoder[PollUpdate] = MainData.dataEncoder
+    override def deserializeUpdate(bytes: Array[Byte]): IO[Either[Throwable, PollUpdate]] = IO(Deserializers.deserializeUpdate(bytes))
 
-      override def dataDecoder: Decoder[PollUpdate] = MainData.dataDecoder
+    override def deserializeBlock(bytes: Array[Byte]): IO[Either[Throwable, Signed[DataApplicationBlock]]] = IO(Deserializers.deserializeBlock(bytes)(dataDecoder.asInstanceOf[Decoder[DataUpdate]]))
```

Comment:
```suggestion
    override def serializeState(state: VoteStateOnChain): IO[Array[Byte]] =
      IO(Serializers.serialize[VoteStateOnChain](state))

    override def serializeUpdate(update: PollUpdate): IO[Array[Byte]] =
      IO(Serializers.serialize[PollUpdate](update))

    override def serializeBlock(block: Signed[DataApplicationBlock]): IO[Array[Byte]] =
      IO {
        implicit val dataUpdateEncoder = dataEncoder.asInstanceOf[Encoder[DataUpdate]]
        Serializers.serialize[DataApplicationBlock](block)
      }

    override def deserializeState(bytes: Array[Byte]): IO[Either[Throwable, VoteStateOnChain]] =
      IO(Deserializers.deserialize[VoteStateOnChain](bytes))

    override def deserializeUpdate(bytes: Array[Byte]): IO[Either[Throwable, PollUpdate]] =
      IO(Deserializers.deserialize[PollUpdate](bytes))

    override def deserializeBlock(bytes: Array[Byte]): IO[Either[Throwable, Signed[DataApplicationBlock]]] =
      IO {
        implicit val dataUpdateDecoder: Decoder[DataUpdate] = dataDecoder.asInstanceOf[Decoder[DataUpdate]]
        Deserializers.deserialize[Signed[DataApplicationBlock]](bytes)
      }
```

Reply:
I prefer to keep with names and call the shared function serialize/deserialize